<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ резюме из HeadHunter
   

In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import re


# Исследование структуры данных

1. Прочитайте данные с помощью библиотеки Pandas. Совет: перед чтением обратите внимание на разделитель внутри файла. 

In [4]:
#ваш код здесь
hh_df = pd.read_csv('data/dst_hh_dtbase.csv', sep=';')
print(f'Строк в датафрейме: {hh_df.shape[0]}')
print(f'Столбцов в датафрейме: {hh_df.shape[1]}')
      
      

Строк в датафрейме: 44744
Столбцов в датафрейме: 12


2. Выведите несколько первых (последних) строк таблицы, чтобы убедиться в том, что ваши данные не повреждены. Ознакомьтесь с признаками и их структурой.

In [ ]:
#ваш код здесь
display(hh_df.head()) #первые 5 строк
display(hh_df.tail()) #последние 5 строк

,"Пол, возраст",ЗП,Ищет работу на должность:,"Город, переезд, командировки",Занятость,График,Опыт работы,Последнее/нынешнее место работы,Последняя/нынешняя должность,Образование и ВУЗ,Обновление резюме,Авто
0,"Мужчина , 39 лет , родился 27 ноября 1979",29000 руб.,Системный администратор,"Советск (Калининградская область) , не готов к...","частичная занятость, проектная работа, полная ...","гибкий график, полный день, сменный график, ва...",Опыт работы 16 лет 10 месяцев Август 2010 — п...,"МАОУ ""СОШ № 1 г.Немана""",Системный администратор,Неоконченное высшее образование 2000 Балтийск...,16.04.2019 15:59,Имеется собственный автомобиль
1,"Мужчина , 60 лет , родился 20 марта 1959",40000 руб.,Технический писатель,"Королев , не готов к переезду , готов к редким...","частичная занятость, проектная работа, полная ...","гибкий график, полный день, сменный график, уд...",Опыт работы 19 лет 5 месяцев Январь 2000 — по...,Временный трудовой коллектив,"Менеджер проекта, Аналитик, Технический писатель",Высшее образование 1981 Военно-космическая ак...,12.04.2019 08:42,Не указано
2,"Женщина , 36 лет , родилась 12 августа 1982",20000 руб.,Оператор,"Тверь , не готова к переезду , не готова к ком...",полная занятость,полный день,Опыт работы 10 лет 3 месяца Октябрь 2004 — Де...,ПАО Сбербанк,Кассир-операционист,Среднее специальное образование 2002 Профессио...,16.04.2019 08:35,Не указано
3,"Мужчина , 38 лет , родился 25 июня 1980",100000 руб.,Веб-разработчик (HTML / CSS / JS / PHP / базы ...,"Саратов , не готов к переезду , готов к редким...","частичная занятость, проектная работа, полная ...","гибкий график, удаленная работа",Опыт работы 18 лет 9 месяцев Август 2017 — Ап...,OpenSoft,Инженер-программист,Высшее образование 2002 Саратовский государст...,08.04.2019 14:23,Не указано
4,"Женщина , 26 лет , родилась 3 марта 1993",140000 руб.,Региональный менеджер по продажам,"Москва , не готова к переезду , готова к коман...",полная занятость,полный день,Опыт работы 5 лет 7 месяцев Региональный мене...,Мармелад,Менеджер по продажам,Высшее образование 2015 Кгу Психологии и педаг...,22.04.2019 10:32,Не указано


,"Пол, возраст",ЗП,Ищет работу на должность:,"Город, переезд, командировки",Занятость,График,Опыт работы,Последнее/нынешнее место работы,Последняя/нынешняя должность,Образование и ВУЗ,Обновление резюме,Авто
44739,"Мужчина , 30 лет , родился 17 января 1989",50000 руб.,"Финансист, аналитик, экономист, бухгалтер, мен...","Тверь , готов к переезду (Москва, Химки) , гот...",полная занятость,"полный день, удаленная работа","Опыт работы 7 лет 7 месяцев Финансист, аналит...","ООО ""IAS"" (независимый участник объединения Ru...",Руководитель субгруппы,Высшее образование 2015 Московский гуманитарн...,22.04.2019 12:32,Не указано
44740,"Мужчина , 27 лет , родился 5 марта 1992",39000 руб.,"Системный администратор, IT-специалист","Липецк , готов к переезду , готов к командировкам","проектная работа, частичная занятость, полная ...","удаленная работа, гибкий график, полный день, ...","Опыт работы 7 лет Системный администратор, IT...",ИП Пестрецов,Предприниматель,Высшее образование (Бакалавр) 2016 Воронежски...,22.04.2019 13:11,Не указано
44741,"Женщина , 48 лет , родилась 26 декабря 1970",40000 руб.,"Аналитик данных, Математик","Челябинск , готова к переезду , готова к редки...",полная занятость,"полный день, удаленная работа",Опыт работы 21 год 5 месяцев Январь 1998 — по...,"ОАО «ЧМК», Исследовательско-Технологический Це...",Начальник группы аналитики,Высшее образование 2000 Южно-Уральский госуда...,09.04.2019 05:07,Не указано
44742,"Мужчина , 24 года , родился 6 октября 1994",20000 руб.,Контент-менеджер,"Тамбов , не готов к переезду , не готов к кома...","частичная занятость, полная занятость",удаленная работа,Опыт работы 3 года 10 месяцев Контент-менедже...,IQ-Maxima,Менеджер проектов,Высшее образование 2015 Тамбовский государств...,26.04.2019 14:25,Имеется собственный автомобиль
44743,"Мужчина , 38 лет , родился 25 апреля 1980",120000 руб.,Руководитель проекта,"Москва , не готов к переезду , не готов к кома...",полная занятость,полный день,Опыт работы 15 лет 10 месяцев Руководитель пр...,ПАО ГК ТНС энерго,Руководитель отдела технической поддержки,Высшее образование 1997 Южно-Российский госуд...,05.07.2018 20:15,Не указано


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44744 entries, 0 to 44743
Data columns (total 12 columns):
 #   Column                           Non-Null Count  Dtype 
---  ------                           --------------  ----- 
 0   Пол, возраст                     44744 non-null  object
 1   ЗП                               44744 non-null  object
 2   Ищет работу на должность:        44744 non-null  object
 3   Город, переезд, командировки     44744 non-null  object
 4   Занятость                        44744 non-null  object
 5   График                           44744 non-null  object
 6   Опыт работы                      44576 non-null  object
 7   Последнее/нынешнее место работы  44743 non-null  object
 8   Последняя/нынешняя должность     44742 non-null  object
 9   Образование и ВУЗ                44744 non-null  object
 10  Обновление резюме                44744 non-null  object
 11  Авто                             44744 non-null  object
dtypes: object(12)
memory usage: 4.1+

None

3. Выведите основную информацию о числе непустых значений в столбцах и их типах в таблице.

4. Обратите внимание на информацию о числе непустых значений.

In [ ]:
#ваш код здесь
display(hh_df.info())
display(hh_df.isnull().sum())
hh_df_copy = hh_df.copy()

5. Выведите основную статистическую информацию о столбцах.


In [12]:
#ваш код здесь
hh_df.describe(include='all')

,"Пол, возраст",ЗП,Ищет работу на должность:,"Город, переезд, командировки",Занятость,График,Опыт работы,Последнее/нынешнее место работы,Последняя/нынешняя должность,Образование и ВУЗ,Обновление резюме,Авто
count,44744,44744,44744,44744,44744,44744,44576,44743,44742,44744,44744,44744
unique,16003,690,14929,10063,38,47,44413,30214,16927,40148,18838,2
top,"Мужчина , 32 года , родился 17 сентября 1986",50000 руб.,Системный администратор,"Москва , не готов к переезду , не готов к кома...",полная занятость,полный день,Опыт работы 10 лет 8 месяцев Апрель 2018 — по...,Индивидуальное предпринимательство / частная п...,Системный администратор,Высшее образование 2014 Московский технически...,07.05.2019 09:50,Не указано
freq,18,4064,3099,1261,30026,22727,3,935,2062,4,25,32268


1. Начнем с простого - с признака **"Образование и ВУЗ"**. Его текущий формат это: **<Уровень образования год выпуска ВУЗ специальность...>**. Например:
* Высшее образование 2016 Московский авиационный институт (национальный исследовательский университет)...
* Неоконченное высшее образование 2000  Балтийская государственная академия рыбопромыслового флота…
Нас будет интересовать только уровень образования.

Создайте с помощью функции-преобразования новый признак **"Образование"**, который должен иметь 4 категории: "высшее", "неоконченное высшее", "среднее специальное" и "среднее".

Выполните преобразование, ответьте на контрольные вопросы и удалите признак "Образование и ВУЗ".

Совет: обратите внимание на структуру текста в столбце **"Образование и ВУЗ"**. Гарантируется, что текущий уровень образования соискателя всегда находится в первых 2ух слов и начинается с заглавной буквы. Воспользуйтесь этим.

*Совет: проверяйте полученные категории, например, с помощью метода unique()*


# Преобразование данных

In [201]:
#ваш код здесь
def educ_level(text):
    first_two_words = ' '.join(text.split()[:2]).lower()
    if 'высшее' in first_two_words and 'неоконченное' not in first_two_words:
        return 'высшее'
    elif 'неоконченное' in first_two_words:
        return 'неоконченное высшее'
    elif 'среднее специальное' in first_two_words:
        return 'среднее специальное'
    elif 'среднее' in first_two_words:
        return 'среднее'
    else:
        return 'не определено'

hh_df['Образование'] = hh_df_copy['Образование и ВУЗ'].apply(educ_level)
#hh_df = hh_df.drop('Образование и ВУЗ', axis=1)

2. Теперь нас интересует столбец **"Пол, возраст"**. Сейчас он представлен в формате **<Пол , возраст , дата рождения >**. Например:
* Мужчина , 39 лет , родился 27 ноября 1979 
* Женщина , 21 год , родилась 13 января 2000
Как вы понимаете, нам необходимо выделить каждый параметр в отдельный столбец.

Создайте два новых признака **"Пол"** и **"Возраст"**. При этом важно учесть:
* Признак пола должен иметь 2 уникальных строковых значения: 'М' - мужчина, 'Ж' - женщина. 
* Признак возраста должен быть представлен целыми числами.

Выполните преобразование, ответьте на контрольные вопросы и удалите признак **"Пол, возраст"** из таблицы.

*Совет: обратите внимание на структуру текста в столбце, в части на то, как разделены параметры пола, возраста и даты рождения между собой - символом ' , '. 
Гарантируется, что структура одинакова для всех строк в таблице. Вы можете воспользоваться этим.*


In [199]:
#ваш код здесь
def gender(text):
    gnd = text.split()[0]
    
    if gnd == 'Мужчина':
        return 'М'
    elif gnd == 'Женщина':
        return 'Ж'
    else:
        return None
def age(text):
    ages = text.split()
    
    for ags in ages:
        if ags.isdigit():  # Нашли первое число (возраст)
            return int(ags)
    return None
hh_df['Пол'] = hh_df_copy['Пол, возраст'].apply(gender)
hh_df['Возраст'] = hh_df_copy['Пол, возраст'].apply(age)
print(hh_df['Пол'].value_counts())
print(hh_df['Возраст'].value_counts())
print(f'Процент женских резюме = {((hh_df['Пол'] == 'Ж').sum() / len(hh_df)) * 100}%')
print(f'Средний возраст соискателей = {hh_df['Возраст'].mean()}')
#hh_df = hh_df.drop('Пол, возраст', axis=1)

Пол
М    36211
Ж     8533
Name: count, dtype: int64
Возраст
30     2834
29     2765
28     2748
31     2675
27     2583
       ... 
15        2
76        1
77        1
14        1
100       1
Name: count, Length: 63, dtype: int64
Процент женских резюме = 19.070713391739673%
Средний возраст соискателей = 32.19674146254246


3. Следующим этапом преобразуем признак **"Опыт работы"**. Его текущий формат - это: **<Опыт работы: n лет m месяцев, периоды работы в различных компаниях…>**. 

Из столбца нам необходимо выделить общий опыт работы соискателя в месяцах, новый признак назовем "Опыт работы (месяц)"

Для начала обсудим условия решения задачи:
* Во-первых, в данном признаке есть пропуски. Условимся, что если мы встречаем пропуск, оставляем его как есть (функция-преобразование возвращает NaN)
* Во-вторых, в данном признаке есть скрытые пропуски. Для некоторых соискателей в столбце стоит значения "Не указано". Их тоже обозначим как NaN (функция-преобразование возвращает NaN)
* В-третьих, нас не интересует информация, которая описывается после указания опыта работы (периоды работы в различных компаниях)
* В-четвертых, у нас есть проблема: опыт работы может быть представлен только в годах или только месяцах. Например, можно встретить следующие варианты:
    * Опыт работы 3 года 2 месяца…
    * Опыт работы 4 года…
    * Опыт работы 11 месяцев…
    * Учитывайте эту особенность в вашем коде

Учитывайте эту особенность в вашем коде

В результате преобразования у вас должен получиться столбец, содержащий информацию о том, сколько месяцев проработал соискатель.
Выполните преобразование, ответьте на контрольные вопросы и удалите столбец **"Опыт работы"** из таблицы.


In [197]:
#ваш код здесь
def exp_work(text):
    if pd.isna(text):
        return np.nan
    
    text_str = str(text).lower()
    exp_match = re.search(r'опыт\s+работы\s+([^\.\n]+)', text_str)
    if not exp_match:
        return np.nan
    
    exp_text = exp_match.group(1)
    
    years = months = 0
    year_match = re.search(r'(\d+)\s*(?:год|года|лет)', exp_text)
    if year_match:
        years = int(year_match.group(1))
    
    month_match = re.search(r'(\d+)\s*(?:месяц|месяца|месяцев)', exp_text)
    if month_match:
        months = int(month_match.group(1))
    
    return years * 12 + months


hh_df['Опыт работы (месяц)'] = hh_df_copy['Опыт работы'].apply(exp_work)

print(f"Максимальный опыт: {hh_df['Опыт работы (месяц)'].max()} месяцев")

print(f"Медиана: {hh_df['Опыт работы (месяц)'].median()} месяцев")
display(hh_df['Опыт работы (месяц)'].describe())
display(hh_df['Опыт работы (месяц)'].head())
#hh_df = hh_df.drop('Опыт работы', axis=1)
display(hh_df.head())

Максимальный опыт: 1188.0 месяцев
Медиана: 100.0 месяцев


count    44574.000000
mean       114.646969
std         79.093640
min          1.000000
25%         57.000000
50%        100.000000
75%        154.000000
max       1188.000000
Name: Опыт работы (месяц), dtype: float64

0    202.0
1    233.0
2    123.0
3    225.0
4     67.0
Name: Опыт работы (месяц), dtype: float64

,Ищет работу на должность:,Последнее/нынешнее место работы,Последняя/нынешняя должность,Обновление резюме,Авто,Образование,Пол,Возраст,Опыт работы (месяц),Город,...,полный день,сменный график,гибкий график,удалённая работа,вахтовый метод,ЗП (руб),Дата обновления,Сумма_зарплаты,Валюта,Валюта_ISO
0,Системный администратор,"МАОУ ""СОШ № 1 г.Немана""",Системный администратор,2019-04-16 15:59:00,Имеется собственный автомобиль,неоконченное высшее,М,39,202.0,другие,...,True,True,True,False,True,29000.0,2019-04-16,29000.0,руб.,RUB
1,Технический писатель,Временный трудовой коллектив,"Менеджер проекта, Аналитик, Технический писатель",2019-04-12 08:42:00,Не указано,высшее,М,60,233.0,другие,...,True,True,True,False,False,40000.0,2019-04-12,40000.0,руб.,RUB
2,Оператор,ПАО Сбербанк,Кассир-операционист,2019-04-16 08:35:00,Не указано,среднее специальное,Ж,36,123.0,другие,...,True,False,False,False,False,20000.0,2019-04-16,20000.0,руб.,RUB
3,Веб-разработчик (HTML / CSS / JS / PHP / базы ...,OpenSoft,Инженер-программист,2019-04-08 14:23:00,Не указано,высшее,М,38,225.0,другие,...,False,False,True,False,False,100000.0,2019-04-08,100000.0,руб.,RUB
4,Региональный менеджер по продажам,Мармелад,Менеджер по продажам,2019-04-22 10:32:00,Не указано,высшее,Ж,26,67.0,Москва,...,True,False,False,False,False,140000.0,2019-04-22,140000.0,руб.,RUB


4. Хорошо идем! Следующий на очереди признак "Город, переезд, командировки". Информация в нем представлена в следующем виде: **<Город , (метро) , готовность к переезду (города для переезда) , готовность к командировкам>**. В скобках указаны необязательные параметры строки. Например, можно встретить следующие варианты:

* Москва , не готов к переезду , готов к командировкам
* Москва , м. Беломорская , не готов к переезду, не готов к командировкам
* Воронеж , готов к переезду (Сочи, Москва, Санкт-Петербург) , готов к командировкам

Создадим отдельные признаки **"Город"**, **"Готовность к переезду"**, **"Готовность к командировкам"**. При этом важно учесть:

* Признак **"Город"** должен содержать только 4 категории: "Москва", "Санкт-Петербург" и "город-миллионник" (их список ниже), остальные обозначьте как "другие".

    Список городов-миллионников:
    
   <code>million_cities = ['Новосибирск', 'Екатеринбург','Нижний Новгород','Казань', 'Челябинск','Омск', 'Самара', 'Ростов-на-Дону', 'Уфа', 'Красноярск', 'Пермь', 'Воронеж','Волгоград']
    </code>
    Инфорация о метро, рядом с которым проживает соискатель нас не интересует.
* Признак **"Готовность к переезду"** должен иметь два возможных варианта: True или False. Обратите внимание, что возможны несколько вариантов описания готовности к переезду в признаке "Город, переезд, командировки". Например:
    * … , готов к переезду , …
    * … , не готова к переезду , …
    * … , готова к переезду (Москва, Санкт-Петербург, Ростов-на-Дону)
    * … , хочу переехать (США) , …
    
    Нас интересует только сам факт возможности или желания переезда.
* Признак **"Готовность к командировкам"** должен иметь два возможных варианта: True или False. Обратите внимание, что возможны несколько вариантов описания готовности к командировкам в признаке "Город, переезд, командировки". Например:
    * … , готов к командировкам , … 
    * … , готова к редким командировкам , …
    * … , не готов к командировкам , …
    
    Нас интересует только сам факт готовности к командировке.
    
    Еще один важный факт: при выгрузки данных у некоторых соискателей "потерялась" информация о готовности к командировкам. Давайте по умолчанию будем считать, что такие соискатели не готовы к командировкам.
    
Выполните преобразования и удалите столбец **"Город, переезд, командировки"** из таблицы.

*Совет: обратите внимание на то, что структура текста может меняться в зависимости от указания ближайшего метро. Учите это, если будете использовать порядок слов в своей программе.*


In [65]:
#ваш код здесь

million_cities = ['Новосибирск', 'Екатеринбург', 'Нижний Новгород', 'Казань', 
                  'Челябинск', 'Омск', 'Самара', 'Ростов-на-Дону', 'Уфа', 
                  'Красноярск', 'Пермь', 'Воронеж', 'Волгоград']

#Функция преобразования в признак Город
def city_of_living(text):

    if pd.isna(text):
        return np.nan
    
    first_part = text.split(' , ')[0].strip()
    
    city = first_part.split(' , ')[0].split(' ,')[0].strip()
  
    if city == 'Москва':
        return 'Москва'
    elif city == 'Санкт-Петербург':
        return 'Санкт-Петербург'
    elif city in million_cities:
        return 'город-миллионник'
    else:
        return 'другие'

hh_df['Город'] = hh_df['Город, переезд, командировки'].apply(city_of_living)

#Функция преобразования в признак 'Готовность к переезду'
def ready_to_relocate(text):
    if pd.isna(text):
        return False
    
    text_lower = str(text).lower()
    
    if 'не готов к переезду' in text_lower or 'не готова к переезду' in text_lower:
        return False
    
    # Проверяем явное указание готовности к переезду (включая варианты со скобками)
    if ('готов к переезду' in text_lower or 
        'готова к переезду' in text_lower or 
        'хочу переехать' in text_lower):
        return True
    
    return False

hh_df['Готовность к переезду'] = hh_df['Город, переезд, командировки'].apply(ready_to_relocate)

#Функция преобразования в признак 'Готовность к командировкам'
import pandas as pd

def ready_to_bsnstrp(text):
    
    if pd.isna(text):
        return False
    
    text_lower = str(text).lower()
    if ('не готов к командировкам' in text_lower or 
        'не готова к командировкам' in text_lower):
        return False
    
    if ('готов к командировкам' in text_lower or 
        'готова к командировкам' in text_lower or
        'готов к редким командировкам' in text_lower or
        'готова к редким командировкам' in text_lower):
        return True
    
    return False

hh_df['Готовность к командировкам'] = hh_df['Город, переезд, командировки'].apply(ready_to_bsnstrp)

#'Город'
print(hh_df['Город'].value_counts())
total = hh_df['Город'].value_counts().sum()
print(f"\nВсего категорий: {hh_df['Город'].nunique()}")
result = round((hh_df['Город'] == 'Санкт-Петербург').mean() * 100, 2)
print(f"В Питере проживает {result}% соискателей")

#'Готовность к переезду'
print(f"\nГотовы к переезду: {(hh_df['Готовность к переезду'] == True).mean()*100:.1f}%")

#'Готовность к командировкам'
print(hh_df['Готовность к командировкам'].value_counts())
print(f"\nГотовы к командировкам: {hh_df['Готовность к командировкам'].mean()*100:.1f}%")

#Готовы и к командировкам и к переезду
both_ready = ((hh_df['Готовность к переезду'] == True) & (hh_df['Готовность к командировкам'] == True)).sum()
print(f'Готовы и к командировкам и на переезд {(both_ready/total) * 100:.1f}% соискателей')
hh_df = hh_df.drop('Город, переезд, командировки', axis=1)

Город
Москва              16621
другие              15854
город-миллионник     7332
Санкт-Петербург      4937
Name: count, dtype: int64

Всего категорий: 4
В Питере проживает 11.03% соискателей

Готовы к переезду: 35.8%
Готовность к командировкам
True     31640
False    13104
Name: count, dtype: int64

Готовы к командировкам: 70.7%
Готовы и к командировкам и на переезд 31.9% соискателей


5. Рассмотрим поближе признаки **"Занятость"** и **"График"**. Сейчас признаки представляют собой набор категорий желаемой занятости (полная занятость, частичная занятость, проектная работа, волонтерство, стажировка) и желаемого графика работы (полный день, сменный график, гибкий график, удаленная работа, вахтовый метод).
На сайте hh.ru соискатель может указывать различные комбинации данных категорий, например:
* полная занятость, частичная занятость
* частичная занятость, проектная работа, волонтерство
* полный день, удаленная работа
* вахтовый метод, гибкий график, удаленная работа, полная занятость

Такой вариант признаков имеет множество различных комбинаций, а значит множество уникальных значений, что мешает анализу. Нужно это исправить!

Давайте создадим признаки-мигалки для каждой категории: если категория присутствует в списке желаемых соискателем, то в столбце на месте строки рассматриваемого соискателя ставится True, иначе - False.

Такой метод преобразования категориальных признаков называется One Hot Encoding и его схема представлена на рисунке ниже:
<img src=https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/ohe.jpg>
Выполните данное преобразование для признаков "Занятость" и "График", ответьте на контрольные вопросы, после чего удалите их из таблицы

In [75]:
#ваш код здесь
busy_categories = ['полная занятость', 'частичная занятость', 'проектная работа', 'волонтерство', 'стажировка']
schedule_categories = ['полный день', 'сменный график', 'гибкий график', 'удалённая работа', 'вахтовый метод']

def sygnal_features(df, column_name, categories_list):
   
    for category in categories_list:
        df[category] = df[column_name].apply(
            lambda x: category in str(x).lower() if pd.notna(x) else False
        )
    return df

hh_df = sygnal_features(hh_df, 'Занятость', busy_categories)
hh_df = sygnal_features(hh_df, 'График', schedule_categories)
employment_columns = [col for col in hh_df.columns if col in busy_categories]
schedule_columns = [col for col in hh_df.columns if col in schedule_categories]
count_p_v = ((hh_df['проектная работа'] == True) & (hh_df['волонтерство'] == True)).sum()
#display(hh_df.head())
count_vahta_flexible = ((hh_df['гибкий график'] == True) & (hh_df['вахтовый метод'] == True)).sum()
print(f"Количество людей, ищущих и проектную работу, и волонтерство: {count_p_v}")
print(f"Количество людей, ищущих и гибкий график, и вахту: {count_vahta_flexible}")
hh_df = hh_df.drop('Занятость', axis=1)
hh_df = hh_df.drop('График', axis=1)



Количество людей, ищущих и проектную работу, и волонтерство: 436
Количество людей, ищущих и гибкий график, и вахту: 2311


6. (2 балла) Наконец, мы добрались до самого главного и самого важного - признака заработной платы **"ЗП"**. 
В чем наша беда? В том, что помимо желаемой заработной платы соискатель указывает валюту, в которой он бы хотел ее получать, например:
* 30000 руб.
* 50000 грн.
* 550 USD

Нам бы хотелось видеть заработную плату в единой валюте, например, в рублях. Возникает вопрос, а где взять курс валют по отношению к рублю?

На самом деле язык Python имеет в арсенале огромное количество возможностей получения данной информации, от обращения к API Центробанка, до использования специальных библиотек, например pycbrf. Однако, это не тема нашего проекта.

Поэтому мы пойдем в лоб: обратимся к специальным интернет-ресурсам для получения данных о курсе в виде текстовых файлов. Например, MDF.RU, данный ресурс позволяет удобно экспортировать данные о курсах различных валют и акций за указанные периоды в виде csv файлов. Мы уже сделали выгрузку курсов валют, которые встречаются в наших данных за период с 29.12.2017 по 05.12.2019. Скачать ее вы можете **на платформе**

Создайте новый DataFrame из полученного файла. В полученной таблице нас будут интересовать столбцы:
* "currency" - наименование валюты в ISO кодировке,
* "date" - дата, 
* "proportion" - пропорция, 
* "close" - цена закрытия (последний зафиксированный курс валюты на указанный день).


Перед вами таблица соответствия наименований иностранных валют в наших данных и их общепринятых сокращений, которые представлены в нашем файле с курсами валют. Пропорция - это число, за сколько единиц валюты указан курс в таблице с курсами. Например, для казахстанского тенге курс на 20.08.2019 составляет 17.197 руб. за 100 тенге, тогда итоговый курс равен - 17.197 / 100 = 0.17197 руб за 1 тенге.
Воспользуйтесь этой информацией в ваших преобразованиях.

<img src=https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/table.jpg>


Осталось только понять, откуда брать дату, по которой определяется курс? А вот же она - в признаке **"Обновление резюме"**, в нем содержится дата и время, когда соискатель выложил текущий вариант своего резюме. Нас интересует только дата, по ней бы и будем сопоставлять курсы валют.

Теперь у нас есть вся необходимая информация для того, чтобы создать признак "ЗП (руб)" - заработная плата в рублях.

После ответа на контрольные вопросы удалите исходный столбец заработной платы "ЗП" и все промежуточные столбцы, если вы их создавали.

Итак, давайте обсудим возможный алгоритм преобразования: 
1. Перевести признак "Обновление резюме" из таблицы с резюме в формат datetime и достать из него дату. В тот же формат привести признак "date" из таблицы с валютами.
2. Выделить из столбца "ЗП" сумму желаемой заработной платы и наименование валюты, в которой она исчисляется. Наименование валюты перевести в стандарт ISO согласно с таблицей выше.
3. Присоединить к таблице с резюме таблицу с курсами по столбцам с датой и названием валюты (подумайте, какой тип объединения надо выбрать, чтобы в таблице с резюме сохранились данные о заработной плате, изначально представленной в рублях). Значение close для рубля заполнить единицей 1 (курс рубля самого к себе)
4. Умножить сумму желаемой заработной платы на присоединенный курс валюты (close) и разделить на пропорцию (обратите внимание на пропуски после объединения в этих столбцах), результат занести в новый столбец "ЗП (руб)".


In [195]:
#ваш код здесь
cur_df = pd.read_csv('data/ExchangeRates.csv', sep=',')
#display(cur_df.info())
#display(cur_df.isnull().sum())
hh_df['Обновление резюме'] = pd.to_datetime(hh_df['Обновление резюме'], errors='coerce')
hh_df['Дата обновления'] = hh_df['Обновление резюме'].dt.date #не забыть удалить
hh_df['Дата обновления'] = pd.to_datetime(hh_df['Дата обновления'], errors='coerce')
cur_df['date'] = pd.to_datetime(cur_df['date'], errors='coerce')
#display(hh_df['Дата обновления'].head())
#display(cur_df['date'].head())
hh_df[['Сумма_зарплаты', 'Валюта']] = hh_df_copy['ЗП'].str.extract(r'(\d+[\s\d]*) (\D+)')
hh_df['Сумма_зарплаты'] = hh_df['Сумма_зарплаты'].str.replace(' ', '').astype(float)
#display(hh_df.head())
cur_map = {
    'руб.': 'RUB',
    'рублей': 'RUB',
    '₽': 'RUB',
    'руб': 'RUB',
    'тенге': 'KZT',
    'KZT': 'KZT',
    'KGS': 'KGS',
    'сом': 'KGS',
    'сомов': 'KGS',
    'USD': 'USD',
    'долларов': 'USD',
    '$': 'USD',
    'евро': 'EUR',
    'EUR': 'EUR',
    '€': 'EUR',
    'бел.руб.': 'BYR',
    'белорусских рублей': 'BYR',
    'грн.': 'UAH',
    'гривен': 'UAH'
}
hh_df['Валюта_ISO'] = hh_df['Валюта'].map(cur_map)
#display(hh_df[['ЗП', 'Сумма_зарплаты', 'Валюта', 'Валюта_ISO']].head(10))

rub = pd.DataFrame([{'currency': 'RUB', 'close': 1.0, 'date': date} 
                       for date in cur_df['date'].unique()])
cur_df = pd.concat([cur_df, rub], ignore_index=True)
# Проблемный кусок кода:
cur_dict = {}
for _, row in cur_df.iterrows():
    key = (row['date'], row['currency'])
    cur_dict[key] = {'close': row['close'], 'proportion': row['proportion']}

for date in cur_df['date'].unique():
    key = (date, 'RUB')
    cur_dict[key] = {'close': 1.0, 'proportion': 1}

def calculate_rub_salary(row):
    key = (row['Дата обновления'], row['Валюта_ISO'])
    if key in cur_dict:
        close = cur_dict[key]['close']
        proportion = cur_dict[key]['proportion']
        return row['Сумма_зарплаты'] * close / proportion
    return row['Сумма_зарплаты']  # Если курс не найден

hh_df['ЗП (руб)'] = hh_df.apply(calculate_rub_salary, axis=1)
#print(hh_df[['Сумма_зарплаты', 'Валюта_ISO', 'ЗП (руб)']].head())
median_salary = round(hh_df['ЗП (руб)'].median(), 2)
print(f'Медианная зарплата: {median_salary}')
#hh_df = hh_df.drop('ЗП', axis=1)

/var/folders/7p/8kmhzc4n149d9rxp80q3n5rr0000gn/T/ipykernel_41374/2124443773.py:8: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  cur_df['date'] = pd.to_datetime(cur_df['date'], errors='coerce')


Медианная зарплата: 58355.73


# Исследование зависимостей в данных

1. Постройте распределение признака **"Возраст"**. Опишите распределение, отвечая на следующие вопросы: чему равна мода распределения, каковы предельные значения признака, в каком примерном интервале находится возраст большинства соискателей? Есть ли аномалии для признака возраста, какие значения вы бы причислили к их числу?
*Совет: постройте гистограмму и коробчатую диаграмму рядом.*

In [256]:
# ваш код здесь
hh_df.info() # Для сверки с тем, что осталось после преобразования
plt.figure(figsize=(15, 6))

# Гистограмма
plt.subplot(1, 2, 1)
plt.hist(hh_df['Возраст'].dropna(), bins=30, edgecolor='black', alpha=0.7, color='skyblue')
plt.title('Распределение возраста соискателей', fontsize=14)
plt.xlabel('Возраст', fontsize=12)
plt.ylabel('Количество соискателей', fontsize=12)
plt.grid(True, alpha=0.3)
med_age = hh_df['Возраст'].median()
plt.axvline(med_age, color='red', linestyle='--', linewidth=2, label=f'Медиана: {med_age:.1f}')
mode_age = hh_df['Возраст'].mode()[0]
plt.axvline(mode_age, color='green', linestyle='--', linewidth=2, label=f'Мода: {mode_age}')
plt.legend()
plt.savefig('plots/AgesDistrib_Hist.png', dpi=300, bbox_inches='tight')
plt.close()

# Коробчатая диаграмма (Boxplot - "Ящик с усами")
plt.subplot(1, 2, 2)
plt.boxplot(hh_df['Возраст'].dropna(), vert=False, patch_artist=True, 
            boxprops=dict(facecolor='lightblue'))
plt.title('Boxplot возраста соискателей', fontsize=14)
plt.xlabel('Возраст', fontsize=12)
plt.grid(True, alpha=0.3)
plt.savefig('plots/AgesDistrib_Boxplot.png', dpi=300, bbox_inches='tight')
plt.close()

# ***Расчеты статистики***:
# Модальное значение возраста
mode_values = hh_df['Возраст'].mode()[0]
print(f'Модальное значение возраста стоискателей: {mode_values}')
# Предельные значения возраста соискателей:
minAge = hh_df['Возраст'].min()
maxAge = hh_df['Возраст'].max()
print(f"Минимальный возраст: {minAge} лет")
print(f"Максимальный возраст: {maxAge} лет")
# Интервал возраста соискателей (68%)
meanAge = hh_df['Возраст'].mean()
stdAge = hh_df['Возраст'].std()
print(f"Интервал возраста большинства соискателей: от {meanAge - stdAge:.0f} до {meanAge + stdAge:.0f} лет")
# Аномалии признака "Возраст". Применю метод 1.5 IQR
Q1 = hh_df['Возраст'].quantile(0.25)
Q3 = hh_df['Возраст'].quantile(0.75)
IQR = Q3 - Q1

lowBoarder = Q1 - 1.5 * IQR
upBoarder = Q3 + 1.5 * IQR

tooLow = hh_df[hh_df['Возраст'] < lowBoarder]['Возраст'].unique()  # Слишком молодые
tooHigh = hh_df[hh_df['Возраст'] > upBoarder]['Возраст'].unique()  # Слишком пожилые

print(f"Границы для нормальных значений: от {lowBoarder:.1f} до {upBoarder:.1f} лет")
print(f"Аномально молодые (<{lowBoarder:.0f} лет): {sorted(tooLow)}")
print(f"Аномально пожилые (>{upBoarder:.0f} лет): {sorted(tooHigh)}")



<class 'pandas.core.frame.DataFrame'>
Index: 44246 entries, 0 to 44743
Data columns (total 23 columns):
 #   Column                           Non-Null Count  Dtype         
---  ------                           --------------  -----         
 0   Ищет работу на должность:        44246 non-null  object        
 1   Последнее/нынешнее место работы  44246 non-null  object        
 2   Последняя/нынешняя должность     44246 non-null  object        
 3   Обновление резюме                44246 non-null  datetime64[ns]
 4   Авто                             44246 non-null  object        
 5   Образование                      44246 non-null  object        
 6   Пол                              44246 non-null  object        
 7   Возраст                          44246 non-null  int64         
 8   Опыт работы (месяц)              44246 non-null  float64       
 9   Город                            44246 non-null  object        
 10  Готовность к переезду            44246 non-null  bool          

ваши выводы по графику здесь

По графику видно следующее: 
- 25-35 лет - возраст 50% соискателей:
- Нижняя граница boxplot (25 лет) = 25-й процентиль (Q1)
- Верхняя граница boxplot (35 лет) = 75-й процентиль (Q3)
- Медиана примерно 30 лет
- "Усы" (14-48 лет) - нормальный диапазон:
    - Левый "ус" (примерно 14 лет) = Q1 - 1.5*IQR
    - Правый "ус" (примерно 48 лет) = Q3 + 1.5*IQR 

Таким образом, всё между усами (14-48 лет) - нормальные значения, а от 48 до 100 лет) - аномалии (выбросы):
- Люди старше 48 лет - по графику аномально пожилые соискатели
- 100 лет - скорее всего какая то ошибка или опечатка в данных

Вывод:

- Нормальный возраст соискателей: 14-48 лет
- Аномалии: люди старше 48 лет (особенно 100 лет)
- Большинство соискателей: 25-35 лет


2. Постройте распределение признака **"Опыт работы (месяц)"**. Опишите данное распределение, отвечая на следующие вопросы: чему равна мода распределения, каковы предельные значения признака, в каком примерном интервале находится опыт работы большинства соискателей? Есть ли аномалии для признака опыта работы, какие значения вы бы причислили к их числу?
*Совет: постройте гистограмму и коробчатую диаграмму рядом.*

In [257]:
# ваш код здесь
#Возможно была ошибка в функции преобразования опыта работы, переписал, применил.
#Ошибка была в том, что я еще в преобразовании убрал аномальное значение 
#Пришлось вернуться и переписать функцию преобразования опыта работы exp_work()
hh_df['Опыт работы (месяц)'] = hh_df_copy['Опыт работы'].apply(exp_work) 

plt.figure(figsize=(15, 6))
#display(hh_df_copy.tail())
#display(hh_df.tail())
# Гистограмма
plt.subplot(1, 2, 1)
plt.hist(hh_df['Опыт работы (месяц)'].dropna(), bins=30, edgecolor='black', alpha=0.7, color='skyblue')
plt.title('Распределение опыта работы соискателей', fontsize=14)
plt.xlabel('Месяцы', fontsize=12)
plt.ylabel('Количество соискателей', fontsize=12)
plt.grid(True, alpha=0.3)
med_experience = hh_df['Опыт работы (месяц)'].median()
plt.axvline(med_experience, color='red', linestyle='--', linewidth=2, label=f'Медиана: {med_experience:.1f}')
mode_experience = hh_df['Опыт работы (месяц)'].mode()[0]
plt.axvline(mode_experience, color='green', linestyle='--', linewidth=2, label=f'Мода: {mode_experience}')
plt.legend()
plt.savefig('plots/WorkExperienceDistrib_Hist.png', dpi=300, bbox_inches='tight')
plt.close()

# Коробчатая диаграмма (Boxplot - "Ящик с усами")
plt.subplot(1, 2, 2)
plt.boxplot(hh_df['Опыт работы (месяц)'].dropna(), vert=False, patch_artist=True, 
            boxprops=dict(facecolor='lightblue'))
plt.title('Boxplot опыта работы соискателей', fontsize=14)
plt.xlabel('Опыт работы (месяц)', fontsize=12)
plt.grid(True, alpha=0.3)
plt.savefig('plots/WorkExperienceDistrib_Boxplot.png', dpi=300, bbox_inches='tight')
plt.close()

# Расчеты статистики:
# Модальное значение опыта работы
mode_values_exp = hh_df['Опыт работы (месяц)'].mode()[0]
print(f'Модальное значение опыта работы стоискателей: {mode_values_exp}')
# Предельные значения опыта работы соискателей:
minExp = hh_df['Опыт работы (месяц)'].min()
maxExp = hh_df['Опыт работы (месяц)'].max()
print(f"Минимальный опыт работы: {minExp} месяцев")
print(f"Максимальный опыт работы: {maxExp} месяцев")
# Интервал опыта работы соискателей (68%)
meanExp = hh_df['Опыт работы (месяц)'].mean()
stdExp = hh_df['Опыт работы (месяц)'].std()
print(f"Интервал опыта работы большинства соискателей: от {meanExp - stdExp:.0f} до {meanExp + stdExp:.0f} месяцев")
# Аномалии признака "Опыт работы (месяц)". Применю метод 1.5 IQR
Q1_Exp = hh_df['Опыт работы (месяц)'].quantile(0.25)
print(f'25-й перцентиль {Q1_Exp}')
Q3_Exp = hh_df['Опыт работы (месяц)'].quantile(0.75)
print(f'75-й перцентиль {Q3_Exp}')
IQR_Exp = Q3_Exp - Q1_Exp

lowBoarderExp = Q1_Exp - 1.5 * IQR_Exp

upBoarderExp = Q3_Exp + 1.5 * IQR_Exp

tooLowExp = hh_df[hh_df['Опыт работы (месяц)'] < lowBoarderExp]['Опыт работы (месяц)'].unique()  # Слишком маленький опыт
tooHighExp = hh_df[hh_df['Опыт работы (месяц)'] > upBoarderExp]['Опыт работы (месяц)'].unique()  # Слишком большой - ошибка

print(f"Границы для нормальных значений: от {lowBoarderExp:.1f} до {upBoarderExp:.1f} месяцев")
print(f"Аномально мало (<{lowBoarderExp:.0f} месяцев): {sorted(tooLowExp)}")
print(f"Аномально большой опыт (>{upBoarderExp:.0f} месяцев): {sorted(tooHighExp)}")

Модальное значение опыта работы стоискателей: 81.0
Минимальный опыт работы: 1.0 месяцев
Максимальный опыт работы: 663.0 месяцев
Интервал опыта работы большинства соискателей: от 36 до 193 месяцев
25-й перцентиль 57.0
75-й перцентиль 154.0
Границы для нормальных значений: от -88.5 до 299.5 месяцев
Аномально мало (<-88 месяцев): []
Аномально большой опыт (>300 месяцев): [np.float64(300.0), np.float64(301.0), np.float64(302.0), np.float64(303.0), np.float64(304.0), np.float64(305.0), np.float64(306.0), np.float64(307.0), np.float64(308.0), np.float64(309.0), np.float64(310.0), np.float64(311.0), np.float64(312.0), np.float64(313.0), np.float64(314.0), np.float64(315.0), np.float64(316.0), np.float64(317.0), np.float64(318.0), np.float64(319.0), np.float64(320.0), np.float64(321.0), np.float64(322.0), np.float64(323.0), np.float64(324.0), np.float64(325.0), np.float64(326.0), np.float64(327.0), np.float64(328.0), np.float64(329.0), np.float64(330.0), np.float64(331.0), np.float64(332.0), n

ваши выводы здесь

По графику видно следующее: 
- Интервал опыта работы большинства соискателей от 36 до 194 месяцев
- Нижняя граница boxplot (57 месяцев) = 25-й процентиль (Q1_Exp)
- Верхняя граница boxplot (154 месяца) = 75-й процентиль (Q3_Exp)
- Медиана примерно 100 месяцев
- "Усы" (-81-300 месяцев) - нормальный диапазон:
    - Левый "ус" (примерно 0 месяцев) = Q1_Exp - 1.5*IQR_Exp
    - Правый "ус" (примерно 300 месяцев) = Q3_Exp + 1.5*IQR_Exp 

Таким образом, всё между "усами" (-81-300 месяцев) - нормальные значения, а от 300 до 1188 лет) - аномалии (выбросы):
- Люди с опытом более 300 месяцев - по графику аномально опытные соискатели
- 1188 месяцев - скорее всего какая то ошибка или опечатка в данных

Вывод:

- Нормальный опыт работы соискателей: 0 - 300 месяцев
- Аномалии: люди с опытом 300 и более месяцев
- 1188 месяцев опыт - ошибка данных
- Большинство соискателей: 36 - 194 месяцев

3. Постройте распределение признака **"ЗП (руб)"**. Опишите данное распределение, отвечая на следующие вопросы: каковы предельные значения признака, в каком примерном интервале находится заработная плата большинства соискателей? Есть ли аномалии для признака возраста? Обратите внимание на гигантские размеры желаемой заработной платы.
*Совет: постройте гистограмму и коробчатую диаграмму рядом.*


In [258]:
# ваш код здесь
#display(hh_df['ЗП (руб)'])
# Я не понимаю почему 22 соискателя не проходят в ответе! Проверяю вручную
hh_df['ЗП (руб)'].to_csv('fd.txt', sep='\t', header=None, index=True, encoding='utf-8')
count = (hh_df['ЗП (руб)'] > 1000000.0).sum()
print(f"Строго > 1 млн: {count}")

salaries = hh_df['ЗП (руб)'].dropna()

# > 1 mln rub
count_moreMillion = (salaries > 1_000_000).sum()

fig, ax = plt.subplots(figsize=(10, 6))

# Гистограмма
counts, bins, patches = ax.hist(salaries / 1000, bins=200, 
                                edgecolor='black', alpha=0.7, color='lightblue')
ax.set_title(f'Гистограмма зарплаты\n{count_moreMillion}  соискателей > 1 млн руб', fontsize=14)
ax.set_xlabel('Зарплата, тысяч рублей', fontsize=12)
ax.set_ylabel('Количество соискателей', fontsize=12)
ax.grid(True, alpha=0.3)
ax.set_ylim(0, 100)
ax.axvline(1000, color='red', linestyle='--', linewidth=1.5, alpha=0.7)
ax.text(1000, 95, f'{count_moreMillion} чел\n> 1 млн руб', 
        ha='center', va='top', color='red', fontsize=12,
        bbox=dict(boxstyle='round,pad=0.3', facecolor='yellow', alpha=0.8))

plt.tight_layout()
plt.savefig('plots/SalaryDistrib_Hist.png', dpi=300, bbox_inches='tight')
plt.close()

print(f"ТОЧНЫЙ РАСЧЁТ:")
print(f"Людей с зарплатой > 1 млн руб: {count_moreMillion}")
print(f"Всего людей: {len(salaries)}")
print(f"Процент: {(count_moreMillion/len(salaries)*100):.4f}%")

# Boxplot
box = ax.boxplot(salaries / 1000, vert=False, patch_artist=True,
                 boxprops=dict(facecolor='lightcoral'),
                 whiskerprops=dict(color='black', linewidth=1.5),
                 capprops=dict(color='black', linewidth=1.5),
                 medianprops=dict(color='yellow', linewidth=2),
                 flierprops=dict(marker='o', markersize=5, alpha=0.5))

ax.set_title(f'Boxplot зарплаты\n{count_moreMillion} человек > 1 млн руб', fontsize=14)
ax.set_xlabel('Зарплата, тысяч рублей', fontsize=12)
ax.grid(True, alpha=0.3)

ax.axvline(1000, color='red', linestyle='--', linewidth=2, alpha=0.8)
ax.text(1000, 1.05, f'{count_moreMillion} чел > 1 млн руб', 
        ha='center', va='bottom', color='red', fontsize=12,
        bbox=dict(boxstyle='round,pad=0.3', facecolor='yellow', alpha=0.8))

median = np.median(salaries / 1000)
q1 = np.percentile(salaries / 1000, 25)
q3 = np.percentile(salaries / 1000, 75)

ax.text(median, 0.7, f'Медиана: {median:.1f}K', 
        ha='center', va='center', fontsize=10)
ax.text(q1, 0.7, f'Q1: {q1:.1f}K', ha='center', va='center', fontsize=10)
ax.text(q3, 0.7, f'Q3: {q3:.1f}K', ha='center', va='center', fontsize=10)

plt.tight_layout()
plt.savefig('plots/SalaryDistrib_Boxplot.png', dpi=300, bbox_inches='tight')
plt.close()

print(f"Людей с зарплатой > 1 млн руб: {count_moreMillion}")

# Предельные значения признака ЗП (руб)
min_salary = hh_df['ЗП (руб)'].min()
max_salary = hh_df['ЗП (руб)'].max()
print(f'Min Salary: {min_salary}')
print(f'Max Salary: {max_salary}')
# Проверка ЗП в 1 руб.
#print(hh_df.iloc[1087]['ЗП (руб)']) 
#print(hh_df_copy.iloc[1087]) - действительно есть такая аномалия

# В каком примерно интервале находится ЗП большинства соискателей:

meanSalary = hh_df['ЗП (руб)'].mean()
stdSalary = hh_df['ЗП (руб)'].std()
medianSalary = round(hh_df['ЗП (руб)'].median(), 2)
modSalary = hh_df['ЗП (руб)'].mode()
round(medianSalary, 2)

print(f"Средняя ЗП: {meanSalary:.0f} руб")
print(f"Стандартное отклонение: {stdSalary:,.0f} руб")
print(f"\n68% соискателей (mean ± std):")
print(f"от {meanSalary - stdSalary:.0f} до {meanSalary + stdSalary:.0f} руб")
print(f'Медианное значение ЗП: {medianSalary} руб.')
print(f'Модальное значение ЗП: {modSalary} руб.')

# Аномалии признака "ЗП (руб)". Применю метод 1.5 IQR
Q1_Sal = hh_df['ЗП (руб)'].quantile(0.25)
print(f'25-й перцентиль {Q1_Sal}')
Q3_Sal = hh_df['ЗП (руб)'].quantile(0.75)
print(f'75-й перцентиль {Q3_Sal}')
IQR_Sal = Q3_Sal - Q1_Sal

lowBoarderSal = Q1_Sal - 1.5 * IQR_Sal

upBoarderSal = Q3_Sal + 1.5 * IQR_Sal

tooLowSal = hh_df[hh_df['ЗП (руб)'] < lowBoarderSal]['ЗП (руб)'].unique()  # Слишком маленькая зп
tooHighSal = hh_df[hh_df['ЗП (руб)'] > upBoarderSal]['ЗП (руб)'].unique()  # Слишком большая зп - ошибка

print(f"Границы для нормальных значений: от {lowBoarderSal:.1f} до {upBoarderSal:.1f} руб.")
print(f"Аномально мало (<{lowBoarderSal:.0f} руб.): {sorted(tooLowSal)}")
print(f"Аномально большая ЗП (>{upBoarderSal:.0f} руб.): {sorted(tooHighSal)}")


Строго > 1 млн: 0
ТОЧНЫЙ РАСЧЁТ:
Людей с зарплатой > 1 млн руб: 0
Всего людей: 44246
Процент: 0.0000%
Людей с зарплатой > 1 млн руб: 0
Min Salary: 1050.0
Max Salary: 923983.0
Средняя ЗП: 75939 руб
Стандартное отклонение: 60,597 руб

68% соискателей (mean ± std):
от 15342 до 136536 руб
Медианное значение ЗП: 60000.0 руб.
Модальное значение ЗП: 0    50000.0
Name: ЗП (руб), dtype: float64 руб.
25-й перцентиль 40000.0
75-й перцентиль 95000.0
Границы для нормальных значений: от -42500.0 до 177500.0 руб.
Аномально мало (<-42500 руб.): []
Аномально большая ЗП (>177500 руб.): [np.float64(178485.75), np.float64(179000.0), np.float64(179808.0), np.float64(179900.0), np.float64(180000.0), np.float64(180277.74999999997), np.float64(180900.50000000003), np.float64(182425.0), np.float64(185000.0), np.float64(185660.80000000002), np.float64(186000.0), np.float64(187000.0), np.float64(187000.31999999998), np.float64(187029.3), np.float64(187099.59), np.float64(190000.0), np.float64(190203.9), np.float

ваши выводы здесь

По графику и расчетам видно следующее: 
- Интервал зарплаты большинства соискателей от -83605 до 239289 руб что весьма странно, скорее всего ошибка в данных
- Нижняя граница (36000 rub) = 25-й процентиль (Q1_Sal)
- Верхняя граница (95000 rub) = 75-й процентиль (Q3_Sal)
- Медиана 58355.73 rub

Ошибочные аномальные данные : -52500 руб. Также и минимум 1 руб тоже скорее всего ошибка данных - в исходных данных такая строка присутствует

Вывод:
- Обычно просят зарплату: 36-95 тыс. руб (половина соискателей)
- Аномалии: зарплаты > 183.5 тыс. руб (включая 22 человека > 1 млн)
- Ошибки данных: отрицательные и крайне низкие значения (1 руб)
- Средняя ЗП испорчена выбросами и поэтому лучше использовать медиану (58356 руб)

4. Постройте диаграмму, которая показывает зависимость **медианной** желаемой заработной платы (**"ЗП (руб)"**) от уровня образования (**"Образование"**). Используйте для диаграммы данные о резюме, где желаемая заработная плата меньше 1 млн рублей.
*Сделайте выводы по представленной диаграмме: для каких уровней образования наблюдаются наибольшие и наименьшие уровни желаемой заработной платы? Как вы считаете, важен ли признак уровня образования при прогнозировании заработной платы?*

In [259]:
# ваш код здесь
#display(hh_df.head())
salLessMillion = hh_df[hh_df['ЗП (руб)'] < 1_000_000]

salByEducation = salLessMillion.groupby('Образование')['ЗП (руб)'].median().sort_values(ascending=False)

# Диаграмма
plt.figure(figsize=(12, 6))
bars = plt.bar(salByEducation.index, salByEducation.values/1000, color='skyblue', edgecolor='black')
plt.title('Зависимость медианной ЗП от Образования\n(ЗП < 1 млн руб)', fontsize=14, pad=20)
plt.xlabel('Уровень образования', fontsize=12)
plt.ylabel('Медианная зарплата, тыс. руб', fontsize=12)
plt.xticks(rotation=45, ha='right')
plt.grid(True, alpha=0.3, axis='y')

for bar in bars:
    height = bar.get_height()
    plt.text(bar.get_x() + bar.get_width()/2., height + 5,
             f'{height:.1f}K', ha='center', va='bottom', fontsize=10)

plt.tight_layout()
plt.savefig('plots/BarOfDependence_SalAndEducation.png', dpi=300, bbox_inches='tight')
plt.close()

maxSalCategoryByEducation = salByEducation.idxmax()
maxSalaryByEducation = salByEducation.max()

print(f"Выше всех оплачивается: {maxSalCategoryByEducation}")
print(f"Медианная зарплата: {maxSalaryByEducation:,.0f} руб ({maxSalaryByEducation/1000:.1f}K руб)")

Выше всех оплачивается: высшее
Медианная зарплата: 60,000 руб (60.0K руб)


ваши выводы здесь

Диаграмма показывает что запрос на зарплату у соискателей с высшим образованием выше. Медиана около 60 тыс руб. 

5. Постройте диаграмму, которая показывает распределение желаемой заработной платы (**"ЗП (руб)"**) в зависимости от города (**"Город"**). Используйте для диаграммы данные о резюме, где желая заработная плата меньше 1 млн рублей.
*Сделайте выводы по полученной диаграмме: как соотносятся медианные уровни желаемой заработной платы и их размах в городах? Как вы считаете, важен ли признак города при прогнозировании заработной платы?*

In [260]:
# ваш код здесь
# Берем топ-10 городов по количеству резюме
top_cities = salLessMillion['Город'].value_counts().head(10).index
filtered_top = salLessMillion[salLessMillion['Город'].isin(top_cities)]

# Диаграмма не влезает, вместо нее scater plot
plt.figure(figsize=(14, 8))
for city in top_cities:
    city_data = filtered_top[filtered_top['Город'] == city]
    plt.scatter(city_data['ЗП (руб)'], [city]*len(city_data), alpha=0.5, s=20, label=city)

plt.title('Распределение ЗП меньше миллиона по городам', fontsize=14)
plt.xlabel('Зарплата, руб', fontsize=12)
plt.ylabel('Город', fontsize=12)
plt.grid(True, alpha=0.3, axis='x')

ax = plt.gca()
ax.xaxis.set_major_formatter(plt.FuncFormatter(lambda x, p: f'{int(x/1000)}K'))

plt.tight_layout()
plt.savefig('plots/ScatterOfDependence_SalAndCity.png', dpi=300, bbox_inches='tight')
plt.close()

salAbou924 = hh_df[(hh_df['ЗП (руб)'] >= 920_000) & (hh_df['ЗП (руб)'] <= 930_000)]

if not salAbou924.empty:
    city = salAbou924.iloc[0]['Город']
    salary = salAbou924.iloc[0]['ЗП (руб)']
    print(f"Город с зарплатой ~924K руб: {city}")
    print(f"Конкретная зарплата: {salary:,.0f} руб")
else:
    print("Зарплат ~924K руб не найдено")

# Статистика:
# Медиана и IRQ
cityStat = hh_df.groupby('Город')['ЗП (руб)'].agg(['median', lambda x: x.quantile(0.75) - x.quantile(0.25)])
cityStat.columns = ['median', 'IQR']
cityStat = cityStat.sort_values('median', ascending=False)

print("Топ-5 городов по медианной зарплате:")
print(cityStat.head())

# Разница между максимальной и минимальной медианой
max_median = cityStat['median'].max()
min_median = cityStat['median'].min()
diffPercent = ((max_median - min_median) / min_median) * 100

print(f"\nРазница медиан между городами:")
print(f"Макс медиана: {max_median:,.0f} руб")
print(f"Мин медиана: {min_median:,.0f} руб")  
print(f"Разница: {diffPercent:.1f}%")

Город с зарплатой ~924K руб: другие
Конкретная зарплата: 923,983 руб
Топ-5 городов по медианной зарплате:
                   median         IQR
Город                                
Москва            85000.0  90000.0000
Санкт-Петербург   60000.0  40000.0000
город-миллионник  40000.0  30000.0000
другие            40000.0  32777.1325

Разница медиан между городами:
Макс медиана: 85,000 руб
Мин медиана: 40,000 руб
Разница: 112.5%


ваши выводы здесь

Выводы по диаграмме:

1. Медианные запросы зарплат различаются между столицами и другими даже крупными городами на 50 - 100 %
2. Размах зарплат (IQR) в крупных городах (Москва, СПб) больше, чем в регионах
3. В городах-миллионниках и городах "другие" медианы примерно равны
4. В регионах зарплаты более сгруппированы, меньше разброс

Признак города важен при прогнозировании:
- Экономическое благополучие и статус города влияют на запросы соискателей

6. Постройте **многоуровневую столбчатую диаграмму**, которая показывает зависимость медианной заработной платы (**"ЗП (руб)"**) от признаков **"Готовность к переезду"** и **"Готовность к командировкам"**. Проанализируйте график, сравнив уровень заработной платы в категориях.

In [261]:
# ваш код здесь
groupByRelocTrip = hh_df.groupby(['Готовность к переезду', 'Готовность к командировкам'])['ЗП (руб)'].median().unstack()

# Диаграмма
x = np.arange(len(groupByRelocTrip.index))
width = 0.35
fig, ax = plt.subplots(figsize=(10, 6))

bars1 = ax.bar(x - width/2, groupByRelocTrip[False]/1000, width, label='Не готов к командировкам', color='lightblue')
bars2 = ax.bar(x + width/2, groupByRelocTrip[True]/1000, width, label='Готов к командировкам', color='coral')

ax.set_title('Медианная зарплата по готовности к переезду и командировкам', fontsize=14)
ax.set_xlabel('Готовность к переезду', fontsize=12)
ax.set_ylabel('Медианная зарплата, тыс. руб', fontsize=12)
ax.set_xticks(x)
ax.set_xticklabels(['Не готов к переезду', 'Готов к переезду'])
ax.legend()
ax.grid(True, alpha=0.3, axis='y')

for bars in [bars1, bars2]:
    for bar in bars:
        height = bar.get_height()
        ax.text(bar.get_x() + bar.get_width()/2., height + 2,
                f'{height:.1f}K', ha='center', va='bottom', fontsize=10)

plt.tight_layout()
plt.savefig('plots/BarOfDependence_SalAndReadyToMove.png', dpi=300, bbox_inches='tight')
plt.close()

readyToRelocBsnsTrip = hh_df[(hh_df['Готовность к переезду'] == True) & 
                 (hh_df['Готовность к командировкам'] == True)]

medianSalary = readyToRelocBsnsTrip['ЗП (руб)'].median()
result = round(medianSalary / 1000)

print(f"Медианная зарплата: {medianSalary:,.0f} руб")
print(f"В тысячах: {result}K руб")

Медианная зарплата: 70,000 руб
В тысячах: 70K руб


ваши выводы здесь

Анализ графика:

1. Готовность к командировкам повышает зарплату примерно на 50-60% в обоих случаях (готов/не готов к переезду)
2. Готовность к переезду дает прирост к запросу ЗП:
  - Готов к переезду: +20-25% к зарплате
  - Готов к командировкам: еще 20-30% к зарплате
3. Максимальная зарплата у тех, кто готов и к переезду, и к командировкам
4. Минимальная зарплата у тех, кто не готов ни к переезду, ни к командировкам
5. Готовность к командировкам дает больший прирост к пожеланиям по ЗП

7. Постройте сводную таблицу, иллюстрирующую зависимость **медианной** желаемой заработной платы от возраста (**"Возраст"**) и образования (**"Образование"**). На полученной сводной таблице постройте **тепловую карту**. Проанализируйте тепловую карту, сравнив показатели внутри групп.

In [262]:
# ваш код здесь

# Таблица
common_table = hh_df.pivot_table(
    values='ЗП (руб)',
    index='Возраст',
    columns='Образование',
    aggfunc='median',
    fill_value=0
)
fin_table = common_table / 1000
print("Сводная таблица Образование, ЗП, Возраст")
print("=" * 70)
print(fin_table.round(1))
print("=" * 70)

# Тепловая карта
plt.figure(figsize=(12, 8))
sns.heatmap(fin_table, cmap='YlOrRd', annot=True, fmt='.1f', 
            linewidths=0.5, linecolor='gray', cbar_kws={'label': 'Зарплата, тыс. руб'})

plt.title('Тепловая карта: Медианная зарплата по возрасту и образованию', fontsize=14, pad=20)
plt.xlabel('Образование', fontsize=12)
plt.ylabel('Возраст', fontsize=12)
plt.xticks(rotation=45, ha='right')
plt.tight_layout()
plt.savefig('plots/WarmMapSalAgesEduc_Dependece.png', dpi=300, bbox_inches='tight')
plt.close()

# Рассчеты
educationSalary = hh_df.groupby('Образование')['ЗП (руб)'].median().sort_values(ascending=False)
print("ЗП и Образование:")
for education, salary in educationSalary.items():
    print(f"{education}: {salary:,.0f} руб ({salary/1000:.1f}K руб)")
print(f"\nКарьерный рост выше: {educationSalary.index[0]}")
print(f"Медианная зарплата: {educationSalary.iloc[0]:,.0f} руб")

Сводная таблица Образование, ЗП, Возраст
Образование  высшее  неоконченное высшее  среднее  среднее специальное
Возраст                                                               
15              0.0                  0.0     10.0                  2.0
16            122.6                  0.0      0.0                 30.0
17             90.1                 80.0     44.0                 22.5
18             30.0                 25.0     30.0                 25.3
19             27.5                 30.0     45.0                 30.0
...             ...                  ...      ...                  ...
72             75.0                  0.0      0.0                  0.0
73             62.5                  0.0      0.0                  0.0
76             45.0                  0.0      0.0                  0.0
77             90.0                  0.0      0.0                  0.0
100             0.0                  0.0      0.0                 60.0

[62 rows x 4 columns]
ЗП и Образова

ваши выводы здесь

Уровень образования является определяющим фактором для роста запроса зарплаты для всех возрастов.

8. Постройте **диаграмму рассеяния**, показывающую зависимость опыта работы (**"Опыт работы (месяц)"**) от возраста (**"Возраст"**). Опыт работы переведите из месяцев в года, чтобы признаки были в едином масштабе. Постройте на графике дополнительно прямую, проходящую через точки (0, 0) и (100, 100). Данная прямая соответствует значениям, когда опыт работы равен возрасту человека. Точки, лежащие на этой прямой и выше нее - аномалии в наших данных (опыт работы больше либо равен возрасту соискателя)

In [263]:
# ваш код здесь

# Переводим опыт из месяцев в годы
hh_df['Опыт работы (годы)'] = hh_df['Опыт работы (месяц)'] / 12

# Диаграмма рассеяния
plt.figure(figsize=(10, 6))
plt.scatter(hh_df['Возраст'], hh_df['Опыт работы (годы)'], 
            alpha=0.5, s=20, color='blue')

plt.title('Зависимость опыта работы от возраста', fontsize=14)
plt.xlabel('Возраст, лет', fontsize=12)
plt.ylabel('Опыт работы, лет', fontsize=12)
plt.grid(True, alpha=0.3)

# Дополнительная прямая (0,0)(100,100) - опыт = возраст
x = np.array([0, 100])
y = x
plt.plot(x, y, 'r-', linewidth=2, label='Опыт = Возраст (аномалия)')

# Область аномалий(выше линии опыт=возраст)
plt.fill_between(x, y, 100, color='red', alpha=0.1, label='Зона аномалий')

plt.legend()
plt.xlim(0, 100)
plt.ylim(0, 100)
plt.tight_layout()
plt.savefig('plots/ScatterPlot_WorkExperience_Ages.png', dpi=300, bbox_inches='tight')
plt.close()

# Считаем аномалии
anomalies = hh_df[hh_df['Опыт работы (годы)'] >= hh_df['Возраст']]
print(f"Найдено аномалий (опыт ≥ возрасту): {len(anomalies)}")
if len(anomalies) > 0:
    print("Примеры аномалий:")
    print(anomalies[['Возраст', 'Опыт работы (годы)']].head())

Найдено аномалий (опыт ≥ возрасту): 0


ваши выводы здесь

**Выводы из графика с аномалиями:**

1. **В целом** :
- Большинство точек лежат ниже линии "опыт = возраст"

2. **Аномалии (7 точек выше линии):**
   - **Опыт работы больше или равен возрасту** - не реально
   - Возможна ошибка в данных

3. **Распределение нормальных данных:**
   - Опыт растет с возрастом линейно
   - Большинство соискателей начинают работать примерно в 20 лет

4. **Качество данных:**
   - 7 аномалий на весь датасет - это немного (<0.1%)
   - Данные в основном качественные
   - Аномалии нужно удалить или исправить

**Дополнительные баллы**

Для получения 2 дополнительных баллов по разведывательному анализу постройте еще два любых содержательных графика или диаграммы, которые помогут проиллюстрировать влияние признаков/взаимосвязь между признаками/распределения признаков. Приведите выводы по ним. Желательно, чтобы в анализе участвовали признаки, которые мы создавали ранее в разделе "Преобразование данных".


In [ ]:
# ваш код здесь

ваши выводы здесь

# Очистка данных

1. Начнем с дубликатов в наших данных. Найдите **полные дубликаты** в таблице с резюме и удалите их. 

In [ ]:
# ваш код здесь
# Непонятно почему ошибка и задание не принимает 158 дубликатов! Пробовал удалить и на преобразованном датафрейме, а теперь на копии
# значение одинаковое
hh_df_noDubl = hh_df.drop_duplicates().reset_index(drop=True)
print(f"Было: {len(hh_df_copy)} записей")
print(f"Стало: {len(hh_df_noDubl)} записей")
print(f"Удалено: {len(hh_df_copy) - len(hh_df_noDubl)} дубликатов")

Найдено полных дубликатов: 0
Было: 44744 записей
Стало: 44586 записей
Удалено: 158 дубликатов


2. Займемся пропусками. Выведите информацию **о числе пропусков** в столбцах. 

In [233]:
# ваш код здесь
missingInfo = hh_df.isnull().sum()
print("Количество пропусков по столбцам:")
for column, count in missingInfo[missingInfo > 0].items():
    print(f"{column}: {count} пропусков ")

missingExp = hh_df['Опыт работы (месяц)'].isnull().sum()
print(f"\nПропусков в столбце 'Опыт работы (месяц)': {missingExp}")


Количество пропусков по столбцам:
Последнее/нынешнее место работы: 1 пропусков 
Последняя/нынешняя должность: 2 пропусков 
Опыт работы (месяц): 168 пропусков 

Пропусков в столбце 'Опыт работы (месяц)': 168


3. Итак, у нас есть пропуски в 3ех столбцах: **"Опыт работы (месяц)"**, **"Последнее/нынешнее место работы"**, **"Последняя/нынешняя должность"**. Поступим следующим образом: удалите строки, где есть пропуск в столбцах с местом работы и должностью. Пропуски в столбце с опытом работы заполните **медианным** значением.

In [242]:
# ваш код здесь
# Понял проблему, забыл удалить столбцы: "Дата обновления", "Сумма_зарплаты", "Валюта" , "Валюта_ISO"
# hh_df = hh_df.drop(columns=['Дата обновления', 'Сумма_зарплаты','Валюта', 'Валюта_ISO']) 

hh_df = hh_df.dropna(subset=['Последнее/нынешнее место работы', 'Последняя/нынешняя должность'])

# Заполняю медианным значением пропуски в столбце "Опыт работы (месяц)"
medianExperience = hh_df['Опыт работы (месяц)'].median()
hh_df['Опыт работы (месяц)'] = hh_df['Опыт работы (месяц)'].fillna(medianExperience)
print(f"Медианный опыт работы: {medianExperience:.1f} месяцев")

# Среднее значение заполненных пропусков опыта работы
meanExperience = hh_df['Опыт работы (месяц)'].mean()
print(f"Среднее значение опыта работы: {meanExperience:.1f} месяцев")


Медианный опыт работы: 100.0 месяцев
Среднее значение опыта работы: 114.6 месяцев


4. Мы добрались до ликвидации выбросов. Сначала очистим данные вручную. Удалите резюме, в которых указана заработная плата либо выше 1 млн. рублей, либо ниже 1 тыс. рублей.

In [ ]:
# ваш код здесь
before = len(hh_df)
print(f"До удаления: {before} строк")
hh_df = hh_df[(hh_df['ЗП (руб)'] < 1_000_000) & (hh_df['ЗП (руб)'] > 1_000)]
after = len(hh_df)
deleted_count = before - after

print(f"После удаления: {before} строк")
print(f"Удалено: {deleted_count} строк")

# Не сохранял копии, теперь напутал и удалил лишнее. Сначала получилось 245, затем удалил зп равно миллиону и тысячи руб.

До удаления: 44339 строк
После удаления: 44253 строк
Удалено: 86 строк


5. В процессе разведывательного анализа мы обнаружили резюме, в которых **опыт работы в годах превышал возраст соискателя**. Найдите такие резюме и удалите их из данных


In [246]:
# ваш код здесь
hh_df['Опыт работы (годы)'] = hh_df['Опыт работы (месяц)'] / 12
anom = hh_df[hh_df['Опыт работы (годы)'] >= hh_df['Возраст']]
print(f"Найдено аномалий (опыт ≥ возрасту): {len(anom)}")
hh_df = hh_df[hh_df['Опыт работы (годы)'] < hh_df['Возраст']]

# Теперь не забуду удалить временный столбец
hh_df = hh_df.drop(columns=['Опыт работы (годы)'])

Найдено аномалий (опыт ≥ возрасту): 7


6. В результате анализа мы обнаружили потенциальные выбросы в признаке **"Возраст"**. Это оказались резюме людей чересчур преклонного возраста для поиска работы. Попробуйте построить распределение признака в **логарифмическом масштабе**. Добавьте к графику линии, отображающие **среднее и границы интервала метода трех сигм**. Напомним, сделать это можно с помощью метода axvline. Например, для построение линии среднего будет иметь вид:

`histplot.axvline(log_age.mean(), color='k', lw=2)`

В какую сторону асимметрично логарифмическое распределение? Напишите об этом в комментарии к графику.
Найдите выбросы с помощью метода z-отклонения и удалите их из данных, используйте логарифмический масштаб. Давайте сделаем послабление на **1 сигму** (возьмите 4 сигмы) в **правую сторону**.

Выведите таблицу с полученными выбросами и оцените, с каким возрастом соискатели попадают под категорию выбросов?

In [264]:
# ваш код здесь
from scipy import stats
logAge = np.log(hh_df['Возраст'])

fig, ax = plt.subplots(figsize=(12, 6))
histPlt = ax.hist(logAge, bins=50, edgecolor='black', alpha=0.7, color='lightblue')
meanLog = logAge.mean()
stdLog = logAge.std()
ax.axvline(meanLog, color='red', lw=3, label=f'Среднее: {np.exp(meanLog):.1f} лет')
# Границы 3 сигм
ax.axvline(meanLog - 3*stdLog, color='orange', linestyle='--', lw=2, 
           label=f'-3-sigma: {np.exp(meanLog - 3*stdLog):.1f} лет')
ax.axvline(meanLog + 3*stdLog, color='orange', linestyle='--', lw=2,
           label=f'+3-sigma: {np.exp(meanLog + 3*stdLog):.1f} лет')

ax.set_title('Распределение возраста (LOG SCALE)', fontsize=14)
ax.set_xlabel('Возраст(log)', fontsize=12)
ax.set_ylabel('Количество', fontsize=12)
ax.grid(True, alpha=0.3)
ax.legend()
plt.savefig('plots/LogHistOfAnomalAges.png', dpi=300, bbox_inches='tight')
plt.close()

# Нормальный не логарифмический масштаб
lowerBound = np.exp(meanLog - 3*stdLog)
upperBound = np.exp(meanLog + 3*stdLog)
print(f"Средний возраст: {np.exp(meanLog):.1f} лет")
print(f"Границы 3-sigma:")
print(f"  Нижняя: {lowerBound:.1f} лет")
print(f"  Верхняя: {upperBound:.1f} лет")

# Считаем выбросы с помощью метода z-отклонения
# Послабление на 1 сигму. Методом 4 сигм в правую сторону
ageMean = hh_df['Возраст'].mean()
ageStd = hh_df['Возраст'].std()

upperBound = ageMean + 4 * ageStd
upperBound_3s = ageMean + 3 * ageStd


# Выбросы (возраст более mean + 4sigma)
outLiers4sigma = hh_df[hh_df['Возраст'] > upperBound]
# Выбросы (возраст более mean + 3sigma)
outLiers3sigma = hh_df[hh_df['Возраст'] > upperBound_3s]

print(f"Метод 4-sigma в правую сторону:")
print(f"Средний возраст: {ageMean:.1f} лет")
print(f"Стандартное отклонение: {ageStd:.1f} лет")
print(f"Верхняя граница (mean + 4-sigma): {upperBound:.1f} лет")
print(f"\nВыбросов 4-sigma (возраст > {upperBound:.0f} лет): {len(outLiers4sigma)}")
print(f"\nВыбросов 3-sigma (возраст > {upperBound_3s:.0f} лет): {len(outLiers3sigma)}")

if len(outLiers4sigma) > 0:
    print(f"\nВозраста выбросов: {sorted(outLiers4sigma['Возраст'].unique())}")
    print(f"\nТаблица выбросов:")
    print(outLiers4sigma[['Возраст', 'ЗП (руб)', 'Город', 'Образование']].to_string())
    
    # Статистика по возрастам выбросов
    print(f"\nСамый младший выброс: {outLiers4sigma['Возраст'].min()} лет")
    print(f"Самый старший выброс: {outLiers4sigma['Возраст'].max()} лет")
    print(f"Средний возраст выбросов: {outLiers4sigma['Возраст'].mean():.1f} лет")

# Подсчет ассиметричности логарифмированной категории "Возраст"
skewness = stats.skew(logAge)
print(f"Коэффициент асимметрии логарифмированного возраста: {skewness:.4f}")
if skewness > 0:
    print("Распределение СКОШЕНО ВПРАВО (положительная асимметрия)")
    print("Хвост вправо - есть выбросы с большими значениями")
elif skewness < 0:
    print("Распределение СКОШЕНО ВЛЕВО (отрицательная асимметрия)")
    print("Хвост влево - есть выбросы с малыми значениями")
else:
    print("Распределение СИММЕТРИЧНО")

Средний возраст: 31.3 лет
Границы 3-sigma:
  Нижняя: 15.7 лет
  Верхняя: 62.6 лет
Метод 4-sigma в правую сторону:
Средний возраст: 32.2 лет
Стандартное отклонение: 7.9 лет
Верхняя граница (mean + 4-sigma): 63.9 лет

Выбросов 4-sigma (возраст > 64 лет): 81

Выбросов 3-sigma (возраст > 56 лет): 628

Возраста выбросов: [np.int64(64), np.int64(65), np.int64(66), np.int64(67), np.int64(68), np.int64(69), np.int64(70), np.int64(71), np.int64(72), np.int64(73), np.int64(76), np.int64(77), np.int64(100)]

Таблица выбросов:
       Возраст  ЗП (руб)             Город          Образование
152         72   25000.0            другие               высшее
850         71  100000.0            Москва               высшее
1430        64   80000.0   Санкт-Петербург               высшее
2106        68   38000.0            Москва               высшее
3271        64   30000.0            Москва               высшее
3293        73   50000.0            Москва               высшее
3460        66   60000.0       

ваш коммментарий здесь

Выводы: 

1. Распределение ассиметрично и скошено вправо  - skewness=0.45 -  больчшая часть это молодежь и средний возраст

2. Возрастной диапазон:

 - Нормальный (3σ): 16-63 года
 - Выбросы по 3σ: возраст > 56 лет (628 человек)
 - Выбросы по 4σ: возраст > 64 лет (81 человек)

3. Выбросы - возрастные соискатели:

- Средний возраст выбросов: 67.3 года
- Наиболее старые > 64 лет (81 человек)
- Возраст > 56 лет уже считается аномалией